src/util/recentlyEditedTree.js

```diff
@@ -0,0 +1,115 @@
+import _ from 'lodash'
+import { timestamp } from './timestamp'
+import { pathToContext } from './pathToContext'
+import { timeDifference } from '../util.js'
+
+const editTimeMax = 30 // time diff limit in second for replacing descendants by ancestor
+
+const findTreeDeepestSubcontext = (tree, path) => {
+  if (path.length === 0) return { node: {}, path: [] }
+  const availableNode = _.get(tree, path)
+  if (availableNode) return { node: availableNode, path }
+  const pathIndex = path.findIndex((value, index) => _.get(tree, (path.slice(0, path.length - index))))
+  return pathIndex > -1 ? { node: _.get(tree, path.slice(0, path.length - pathIndex)), path: path.slice(0, path.length - pathIndex) } : {}
+}
+
+export const findTreeDescendants = (tree, startingPath, leafNodes = []) => {
+  const node = _.get(tree, startingPath)
+  if (!node) return []
+  if (node.leaf) leafNodes.push({ ...node }) // eslint-disable-line fp/no-mutating-methods
+  else Object.keys(node).forEach(child => findTreeDescendants(tree, [...startingPath, child], leafNodes))
+  return leafNodes
+}
+
+const findClosestAncestorWithMultipleChildren = (tree, context) => {
+  const contextIndex = context.findIndex((value, index) => {
+    const node = _.get(tree, (context.slice(0, context.length - index)))
+    return !node.leaf && Object.keys(node).length > 1
+  })
```

Comment:
Similar to `findTreeDeepestSubcontext`, this approach has a lot of overhead due to calling `_.get` in every iteration of `context.findIndex`. Instead, traverse manually. 

`_.get` is great for one-off deep object access, but not as much for multiple adjacent lookups, i.e.:

```
_.get(tree, 'a.b.c.d.e')
_.get(tree, 'a.b.c.d')
_.get(tree, 'a.b.c')
_.get(tree, 'a.b')
_.get(tree, 'a')
```

In this case `_.get` is doing the same work over and over again unnecessarily.

Reply:
Yes I agree. We should traverse manually.